src/rtps/UART_node.cxx

```diff
@@ -0,0 +1,318 @@
+    tc.c_cflag &= ~(CSIZE | PARENB | CBAUD | CRTSCTS);
+    tc.c_cflag |= CS8 | BOTHER;
+
+    tc.c_cc[VMIN] = 0;
+    tc.c_cc[VTIME] = 0;
+    tc.c_ispeed = baudrate;
+    tc.c_ospeed = baudrate;
+
+    if (ioctl(fd, TCSETS2, &tc) == -1) {
+        log_error("Could not set terminal attributes (%m)");
+        goto fail;
+    }
+
+    if (ioctl(fd, TCFLSH, TCIOFLUSH) == -1) {
+        log_error("Could not flush terminal (%m)");
+        goto fail;
+    }
+
+    log_info("Open UART [%d] %s:%u *\n", fd, uart_name, baudrate);
+
+    m_uart_filestream = fd;
+    return m_uart_filestream;
+
+fail:
+    ::close(fd);
+    fd = -1;
+    return -1;
+}
+
+uint8_t UART_node::close_uart()
+{
+    log_debug("Close UART\n");
+    close(m_uart_filestream);
+    return 0;
+}
+
+int UART_node::read()
+{
+    /* Discard whole buffer if it's filled beyond a threshold,
+     * This should prevent buffer being filled by garbage that
+     * no reader (MAVLink or RTPS) can understand.
+     *
+     * TODO: a better approach would be checking if both reader
+     * start understanding messages beyond a certain buffer size,
+     * meaning that everything before is garbage.
+     */
+    if (_buf_size > BUFFER_THRESHOLD) {
+        _buf_size = 0;
+    }
+
+    int r = ::read(m_uart_filestream, _buf + _buf_size, sizeof(_buf) - _buf_size);
+    if (r < 0) {
+        if (errno != EAGAIN) {
+            log_error("Error reading UART %d\n", errno);
+        }
+        return r;
+    }
+
+    _buf_size += r;
+
+    return r;
+}
+
+int UART_node::parseMavlinkFromUART(char buffer[], size_t buflen)
+{
+    unsigned i;
+
+    if (m_uart_filestream == -1)
+        return 2;
+
+    if (_buf_size < 3)
+        return 0;
+
+    // Search for a mavlink packet on buffer to send it
+    i = 0;
+    while (i < (_buf_size - 3) && _buf[i] != 253 && _buf[i] != 254)
+        i++;
+
+    // We need at least the first three bytes to get packet len
+    if (i == _buf_size - 3) {
+        return 0;
+    }
+
+    uint16_t packet_len;
+    if (_buf[i] == 253) {
+        uint8_t payload_len = _buf[i + 1];
+        uint8_t incompat_flags = _buf[i + 2];
+        packet_len = payload_len + 12;
+
+        if (incompat_flags & 0x1) { // signing
+            packet_len += 13;
+        }
+    } else {
+        packet_len = _buf[i + 1] + 8;
+    }
+
+    // cerr << "mavlink parse. message len " << (int)packet_len << " on: " << i << endl;
+
+    // packet is bigger than what we've read, better luck next time
+    if (i + packet_len > _buf_size) {
+        return 0;
+    }
+
+    // buffer should be big enough to hold a mavlink packet
+    if (packet_len > buflen) {
+        return -EMSGSIZE;
+    }
+
+    // Found a whole message, send it and remove from local _buf
+    memmove(buffer, _buf + i, packet_len);
+    memmove(_buf + i, _buf + i + packet_len, sizeof(_buf) - i - packet_len);
+    _buf_size -= packet_len;
+
+    return packet_len;
```

Comment:
This looks like a lot of repetition from code we have in the repo. And it's actually from earlier versions of it.
Besides that, I'm surprised by the copyright on this file since it looks so similar to what we had on our repo already.  Maybe license was wrongly copied and pasted?